hypothesis-python/src/hypothesis/internal/conjecture/data.py

```diff
@@ -159,6 +161,177 @@ def trivial(self):
         return self.forced or self.all_zero
 
 
+def calc_bits_to_array_codes():
+    """Return a list of the smallest array codes that can be used to
+    represent an unsigned integer of size n, for n from 0 to 64 inclusive.
+    """
+    code_iter = iter(["B", "H", "I", "L", "Q"])
+    result = []
+    code = next(code_iter)
+    while len(result) < 65:
+        trial_number = (1 << len(result)) - 1
+        assert trial_number.bit_length() == len(result)
+        try:
+            array_or_list(code, [trial_number])
+            result.append(code)
+        except OverflowError:
+            code = next(code_iter)
+    return result
+
+
+BIT_LENGTH_TO_ARRAY_CODES = calc_bits_to_array_codes()
+
+
+class Blocks(object):
+    """A lazily calculated list of blocks for a particular ``ConjectureResult``
+    or ``ConjectureData`` object.
+
+    Pretends to be a list containing ``Block`` objects but actually only
+    contains their endpoints right up until the point where you want to
+    access the actual block, at which point it is constructed.
+
+    This is designed to be as space efficient as possible, so will at
+    various points silently transform its representation into one
+    that is better suited for the current access pattern.
+
+    In addition, it has a number of convenience methods for accessing
+    properties of the block object at index ``i`` that should generally
+    be preferred to using the Block objects directly, as it will not
+    have to allocate the actual object."""
+
+    __slots__ = ("__endpoints", "owner", "__blocks", "__count", "__sparse")
+
+    def __init__(self, owner):
+        self.owner = owner
+        self.__endpoints = array_or_list("B", [])
+        self.__blocks = {}
+        self.__count = 0
+        self.__sparse = True
+
+    def add_endpoint(self, n):
+        """Add n to the list of enpoints."""
+        assert isinstance(self.owner, ConjectureData)
+
+        # We store endpoints as the smallest size integer we can get
+        # away with. When we try to add an endpoint that is too large,
+        # we upgrade the array to the smallest word size needed to store
+        # the new result.
+        try:
+            self.__endpoints.append(n)
+        except OverflowError:
+            self.__endpoints = array_or_list(
+                BIT_LENGTH_TO_ARRAY_CODES[n.bit_length()], self.__endpoints
```

Comment:
Instead of the complicated `BIT_LENGTH_TO_ARRAY_CODES` calculation, I think all we need is to know which code comes next:

```python
try:
    self.__endpoints.append(n)
except OverflowError:
    # Bump array to larger uint type; does not handle uint64 overflow.
    self.__endpoints = array_or_list(
        dict(B="H", H="I", I="L", L="Q")[self.__endpoints.typecode],
        self.__endpoints
    )
    # Not raw append because we could overflow two small endpoints at once.
    self.add_endpoint(n)
```

Reply:
This doesn't seem significantly less complicated and moves a non-trivial amount of computation to be recomputed each time it's used during the hot path? (It's also incorrect - it assumes we never skip over an interim size, which we could do given a large enough block draw)

Reply:
Oh, sorry, I see you did account for the second bit by recursing. My bad. Anyway, I've switched it over to something based on this idea, and also factored it out into an `IntList` class to keep it a bit more contained.